### ISSUE TITLE
Error in `passenger start` 5.0.X (standalone)

### ISSUE THREAD

**Issue by mrbongiolo** [phusion/passenger#1719]
I first encountered this problem when trying to build a Roda app. I thought it was something with my setup in there, but even with a minimal rack app it kept on crashing.

Here is the log:

```
App 7172 stdout:
App 7172 stderr: /home/vagrant/.gem/ruby/2.2.3/gems/bundler-1.10.6/lib/bundler/spec_set.rb:92:in `block in materialize': Could not find rack-1.6.4 in any of the sources (Bundler::GemNotFound)
App 7172 stderr:        from /home/vagrant/.gem/ruby/2.2.3/gems/bundler-1.10.6/lib/bundler/spec_set.rb:85:in `map!'
App 7172 stderr:        from /home/vagrant/.gem/ruby/2.2.3/gems/bundler-1.10.6/lib/bundler/spec_set.rb:85:in `materialize'
App 7172 stderr:        from /home/vagrant/.gem/ruby/2.2.3/gems/bundler-1.10.6/lib/bundler/definition.rb:140:in `specs'
App 7172 stderr:        from /home/vagrant/.gem/ruby/2.2.3/gems/bundler-1.10.6/lib/bundler/definition.rb:185:in `specs_for'
App 7172 stderr:        from /home/vagrant/.gem/ruby/2.2.3/gems/bundler-1.10.6/lib/bundler/definition.rb:174:in `requested_specs'
App 7172 stderr:        from /home/vagrant/.gem/ruby/2.2.3/gems/bundler-1.10.6/lib/bundler/environment.rb:18:in `requested_specs'
App 7172 stderr:        from /home/vagrant/.gem/ruby/2.2.3/gems/bundler-1.10.6/lib/bundler/runtime.rb:13:in `setup'
App 7172 stderr:        from /home/vagrant/.gem/ruby/2.2.3/gems/bundler-1.10.6/lib/bundler.rb:127:in `setup'
App 7172 stderr:        from /home/vagrant/.gem/ruby/2.2.3/gems/bundler-1.10.6/lib/bundler/setup.rb:18:in `<top (required)>'
App 7172 stderr:        from /home/vagrant/.rubies/ruby-2.2.3/lib/ruby/2.2.0/rubygems/core_ext/kernel_require.rb:54:in `require'
App 7172 stderr:        from /home/vagrant/.rubies/ruby-2.2.3/lib/ruby/2.2.0/rubygems/core_ext/kernel_require.rb:54:in `require'
[ 2016-01-08 15:50:35.8546 7129/7f076ce4a700 age/Cor/App/Implementation.cpp:304 ]: Could not spawn process for application /home/vagrant/development/passenger-failing: An error occurred while starting up the preloader. It exited before signalling successful startup back to Phusion Passenger.
  Error ID: 1b64e7ed
  Error details saved to: /tmp/passenger-error-WpMf4m.html
  Message from application: An error occurred while starting up the preloader. It exited before signalling successful startup back to Phusion Passenger. Please read <a href="https://github.com/phusion/passenger/wiki/Debugging-application-startup-problems">this article</a> for more information about this problem.<br>
<h2>Raw process output:</h2>
<pre>
/home/vagrant/.gem/ruby/2.2.3/gems/bundler-1.10.6/lib/bundler/spec_set.rb:92:in `block in materialize&apos;: Could not find rack-1.6.4 in any of the sources (Bundler::GemNotFound)
        from /home/vagrant/.gem/ruby/2.2.3/gems/bundler-1.10.6/lib/bundler/spec_set.rb:85:in `map!&apos;
        from /home/vagrant/.gem/ruby/2.2.3/gems/bundler-1.10.6/lib/bundler/spec_set.rb:85:in `materialize&apos;
        from /home/vagrant/.gem/ruby/2.2.3/gems/bundler-1.10.6/lib/bundler/definition.rb:140:in `specs&apos;
        from /home/vagrant/.gem/ruby/2.2.3/gems/bundler-1.10.6/lib/bundler/definition.rb:185:in `specs_for&apos;
        from /home/vagrant/.gem/ruby/2.2.3/gems/bundler-1.10.6/lib/bundler/definition.rb:174:in `requested_specs&apos;
        from /home/vagrant/.gem/ruby/2.2.3/gems/bundler-1.10.6/lib/bundler/environment.rb:18:in `requested_specs&apos;
        from /home/vagrant/.gem/ruby/2.2.3/gems/bundler-1.10.6/lib/bundler/runtime.rb:13:in `setup&apos;
        from /home/vagrant/.gem/ruby/2.2.3/gems/bundler-1.10.6/lib/bundler.rb:127:in `setup&apos;
        from /home/vagrant/.gem/ruby/2.2.3/gems/bundler-1.10.6/lib/bundler/setup.rb:18:in `&lt;top (required)&gt;&apos;
        from /home/vagrant/.rubies/ruby-2.2.3/lib/ruby/2.2.0/rubygems/core_ext/kernel_require.rb:54:in `require&apos;
        from /home/vagrant/.rubies/ruby-2.2.3/lib/ruby/2.2.0/rubygems/core_ext/kernel_require.rb:54:in `require&apos;
</pre>

[ 2016-01-08 15:50:35.8610 7129/7f076af9d700 age/Cor/Con/CheckoutSession.cpp:277 ]: [Client 1-1] Cannot checkout session because a spawning error occurred. The identifier of the error is 1b64e7ed. Please see earlier logs for details about the error.
```

My Gemfile
```
source 'https://rubygems.org'
ruby '2.2.3'

gem 'rack'

# 4.0.x works, but 5.0.x not =/
gem 'passenger'#, '~> 4.0'
```

the config.ru
```
require_relative 'app'

run App
```

and the app.rb
```
require 'rack'

class App
  def self.call(env)
    ['200', {'Content-Type' => 'text/html'}, ['A barebones rack app.']]
  end
end
```

Nothing fancy in there.

The 4.0.X standalone version works fine, it seems like the env or gems path are being messed up in some way. Is there anything that I should setup differently when using 5? Couldn't find anything special in the upgrade docs.

**Comment by OnixGH** [comment]
Do you have `rack-1.6.4` installed in your 2.2.3 gems?

**Comment by mrbongiolo** [comment]
Yes, I do.

Check the HTML log file I found this:
```
GREP_COLOR=4;33
NVM_RC_VERSION=
XDG_SESSION_ID=26
LESS_TERMCAP_mb=[04;33m
LESS_TERMCAP_md=[04;33m
GEM_HOME=/home/vagrant/local/ruby/gems
LESS_TERMCAP_me=[0m
SHELL=/bin/bash
TERM=xterm
HISTSIZE=1000000
SSH_CLIENT=10.0.2.2 63151 22
INSTALL_DIR=/home/vagrant/local
PASSENGER_DEBUG_DIR=/tmp/passenger.spawn-debug.XXXXSqm4Zy
OLDPWD=/home/vagrant/development/passenger-failing
LESS_TERMCAP_ue=[0m
SSH_TTY=/dev/pts/0
LC_ALL=en_US.UTF-8
CDPATH=.:/vagrant:/Projects:/home/vagrant:/home/vagrant/Projects
USER=vagrant
NVM_DIR=/home/vagrant/.nvm
LS_COLORS=rs=0:di=01;34:ln=01;36:mh=00:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:su=37;41:sg=30;43:ca=30;41:tw=30;42:ow=34;42:st=37;44:ex=01;32:*.tar=01;31:*.tgz=01;31:*.arj=01;31:*.taz=01;31:*.lzh=01;31:*.lzma=01;31:*.tlz=01;31:*.txz=01;31:*.zip=01;31:*.z=01;31:*.Z=01;31:*.dz=01;31:*.gz=01;31:*.lz=01;31:*.xz=01;31:*.bz2=01;31:*.bz=01;31:*.tbz=01;31:*.tbz2=01;31:*.tz=01;31:*.deb=01;31:*.rpm=01;31:*.jar=01;31:*.war=01;31:*.ear=01;31:*.sar=01;31:*.rar=01;31:*.ace=01;31:*.zoo=01;31:*.cpio=01;31:*.7z=01;31:*.rz=01;31:*.jpg=01;35:*.jpeg=01;35:*.gif=01;35:*.bmp=01;35:*.pbm=01;35:*.pgm=01;35:*.ppm=01;35:*.tga=01;35:*.xbm=01;35:*.xpm=01;35:*.tif=01;35:*.tiff=01;35:*.png=01;35:*.svg=01;35:*.svgz=01;35:*.mng=01;35:*.pcx=01;35:*.mov=01;35:*.mpg=01;35:*.mpeg=01;35:*.m2v=01;35:*.mkv=01;35:*.webm=01;35:*.ogm=01;35:*.mp4=01;35:*.m4v=01;35:*.mp4v=01;35:*.vob=01;35:*.qt=01;35:*.nuv=01;35:*.wmv=01;35:*.asf=01;35:*.rm=01;35:*.rmvb=01;35:*.flc=01;35:*.avi=01;35:*.fli=01;35:*.flv=01;35:*.gl=01;35:*.dl=01;35:*.xcf=01;35:*.xwd=01;35:*.yuv=01;35:*.cgm=01;35:*.emf=01;35:*.axv=01;35:*.anx=01;35:*.ogv=01;35:*.ogx=01;35:*.aac=00;36:*.au=00;36:*.flac=00;36:*.mid=00;36:*.midi=00;36:*.mka=00;36:*.mp3=00;36:*.mpc=00;36:*.ogg=00;36:*.ra=00;36:*.wav=00;36:*.axa=00;36:*.oga=00;36:*.spx=00;36:*.xspf=00;36:
GIT_AUTHOR_NAME=Ralf Schmitz Bongiolo
PAGER=less
PASSENGER_USE_FEEDBACK_FD=true
LESS_TERMCAP_us=[00;32m
PATH=/home/vagrant/local/ruby/gems/bin:/home/vagrant/local/node/npm/bin:/home/vagrant/local/bin:/home/vagrant/.bash/bin:./node_modules/.bin:./bin:/usr/local/bin:/usr/local/sbin:/home/vagrant/.gem/ruby/2.2.3/bin:/usr/local/heroku/bin:/home/vagrant/.rubies/ruby-2.2.3/lib/ruby/gems/2.2.0/bin:/home/vagrant/.rubies/ruby-2.2.3/bin:/home/vagrant/local/ruby/gems/bin:/home/vagrant/local/node/npm/bin:/home/vagrant/local/bin:/home/vagrant/.bash/bin:./node_modules/.bin:./bin:/usr/local/bin:/usr/local/sbin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games
MAIL=/var/mail/vagrant
NVM_NODEJS_ORG_MIRROR=https://nodejs.org/dist
PWD=/home/vagrant/development/passenger-failing
EDITOR=vim
CDHISTORY=/tmp/cd-vagrant
LANG=en_US
RUBY_ENGINE=ruby
HISTCONTROL=ignoreboth:erasedups
HOME=/home/vagrant
SHLVL=1
BUNDLE_GEMFILE=/home/vagrant/development/passenger-failing/Gemfile
GREP_OPTIONS=--color=auto
GEM_ROOT=/home/vagrant/.rubies/ruby-2.2.3/lib/ruby/gems/2.2.0
SERVER_SOFTWARE=nginx/1.8.0 Phusion_Passenger/5.0.23
LOGNAME=vagrant
LESS=-REX
VISUAL=vim
SSH_CONNECTION=10.0.2.2 63151 10.0.2.15 22
GEM_PATH=/home/vagrant/local/ruby/gems
LESS_TERMCAP_so=[01;44;33m
LESSOPEN=| /usr/bin/lesspipe %s
RUBY_ROOT=/home/vagrant/.rubies/ruby-2.2.3
NVM_IOJS_ORG_MIRROR=https://iojs.org/dist
XDG_RUNTIME_DIR=/run/user/1000
GIT_AUTHOR_EMAIL=ralfsb2008@gmail.com
RUBYOPT=-rbundler/setup 
BUNDLE_BIN_PATH=/home/vagrant/.gem/ruby/2.2.3/gems/bundler-1.10.6/bin/bundle
_ORIGINAL_GEM_PATH=/home/vagrant/.gem/ruby/2.2.3:/home/vagrant/.rubies/ruby-2.2.3/lib/ruby/gems/2.2.0:/home/vagrant/local/ruby/gems
RUBYLIB=/home/vagrant/.gem/ruby/2.2.3/gems/bundler-1.10.6/lib
LESSCLOSE=/usr/bin/lesspipe %s %s
RUBY_VERSION=2.2.3
HISTFILE=/home/vagrant/.bash_history
NPM_HOME=/home/vagrant/local/node/npm
LESS_TERMCAP_se=[0m
IN_PASSENGER=1
PYTHONUNBUFFERED=1
NODE_PATH=/home/vagrant/.gem/ruby/2.2.3/gems/passenger-5.0.23/src/nodejs_supportlib
RAILS_ENV=development
RACK_ENV=development
WSGI_ENV=development
NODE_ENV=development
PASSENGER_APP_ENV=development
```

The GEM_HOME and GEM_PATH are not pointing to the correct folder. I'm using `chruby`, but never had a problem with 4.0. 

The bundle exec env output is this:
```
NVM_RC_VERSION=
GREP_COLOR=4;33
LESS_TERMCAP_mb=
XDG_SESSION_ID=27
LESS_TERMCAP_md=
LESS_TERMCAP_me=
GEM_HOME=/home/vagrant/.gem/ruby/2.2.3
TERM=xterm
SHELL=/bin/bash
HISTSIZE=1000
SSH_CLIENT=10.0.2.2 63441 22
INSTALL_DIR=/home/vagrant/local
OLDPWD=/home/vagrant/development
LESS_TERMCAP_ue=
SSH_TTY=/dev/pts/2
LC_ALL=en_US.UTF-8
CDPATH=.:/vagrant:/Projects:/home/vagrant:/home/vagrant/Projects
NVM_DIR=/home/vagrant/.nvm
USER=vagrant
LS_COLORS=rs=0:di=01;34:ln=01;36:mh=00:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:su=37;41:sg=30;43:ca=30;41:tw=30;42:ow=34;42:st=37;44:ex=01;32:*.tar=01;31:*.tgz=01;31:*.arj=01;31:*.taz=01;31:*.lzh=01;31:*.lzma=01;31:*.tlz=01;31:*.txz=01;31:*.zip=01;31:*.z=01;31:*.Z=01;31:*.dz=01;31:*.gz=01;31:*.lz=01;31:*.xz=01;31:*.bz2=01;31:*.bz=01;31:*.tbz=01;31:*.tbz2=01;31:*.tz=01;31:*.deb=01;31:*.rpm=01;31:*.jar=01;31:*.war=01;31:*.ear=01;31:*.sar=01;31:*.rar=01;31:*.ace=01;31:*.zoo=01;31:*.cpio=01;31:*.7z=01;31:*.rz=01;31:*.jpg=01;35:*.jpeg=01;35:*.gif=01;35:*.bmp=01;35:*.pbm=01;35:*.pgm=01;35:*.ppm=01;35:*.tga=01;35:*.xbm=01;35:*.xpm=01;35:*.tif=01;35:*.tiff=01;35:*.png=01;35:*.svg=01;35:*.svgz=01;35:*.mng=01;35:*.pcx=01;35:*.mov=01;35:*.mpg=01;35:*.mpeg=01;35:*.m2v=01;35:*.mkv=01;35:*.webm=01;35:*.ogm=01;35:*.mp4=01;35:*.m4v=01;35:*.mp4v=01;35:*.vob=01;35:*.qt=01;35:*.nuv=01;35:*.wmv=01;35:*.asf=01;35:*.rm=01;35:*.rmvb=01;35:*.flc=01;35:*.avi=01;35:*.fli=01;35:*.flv=01;35:*.gl=01;35:*.dl=01;35:*.xcf=01;35:*.xwd=01;35:*.yuv=01;35:*.cgm=01;35:*.emf=01;35:*.axv=01;35:*.anx=01;35:*.ogv=01;35:*.ogx=01;35:*.aac=00;36:*.au=00;36:*.flac=00;36:*.mid=00;36:*.midi=00;36:*.mka=00;36:*.mp3=00;36:*.mpc=00;36:*.ogg=00;36:*.ra=00;36:*.wav=00;36:*.axa=00;36:*.oga=00;36:*.spx=00;36:*.xspf=00;36:
GIT_AUTHOR_NAME=Ralf Schmitz Bongiolo
PAGER=less
LESS_TERMCAP_us=
MAIL=/var/mail/vagrant
PATH=/home/vagrant/.gem/ruby/2.2.3/bin:/usr/local/heroku/bin:/home/vagrant/.rubies/ruby-2.2.3/lib/ruby/gems/2.2.0/bin:/home/vagrant/.rubies/ruby-2.2.3/bin:/home/vagrant/local/ruby/gems/bin:/home/vagrant/local/node/npm/bin:/home/vagrant/local/bin:/home/vagrant/.bash/bin:./node_modules/.bin:./bin:/usr/local/bin:/usr/local/sbin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games
NVM_NODEJS_ORG_MIRROR=https://nodejs.org/dist
PWD=/home/vagrant/development/team-map
CDHISTORY=/tmp/cd-vagrant
EDITOR=vim
LANG=en_US
RUBY_ENGINE=ruby
HISTCONTROL=ignoreboth
SHLVL=1
HOME=/home/vagrant
GREP_OPTIONS=--color=auto
GEM_ROOT=/home/vagrant/.rubies/ruby-2.2.3/lib/ruby/gems/2.2.0
LESS=-REX
LOGNAME=vagrant
VISUAL=vim
LESS_TERMCAP_so=
GEM_PATH=/home/vagrant/.gem/ruby/2.2.3:/home/vagrant/.rubies/ruby-2.2.3/lib/ruby/gems/2.2.0:/home/vagrant/local/ruby/gems
SSH_CONNECTION=10.0.2.2 63441 10.0.2.15 22
RUBY_ROOT=/home/vagrant/.rubies/ruby-2.2.3
LESSOPEN=| /usr/bin/lesspipe %s
NVM_IOJS_ORG_MIRROR=https://iojs.org/dist
XDG_RUNTIME_DIR=/run/user/1000
RUBYOPT=-rbundler/setup
GIT_AUTHOR_EMAIL=ralfsb2008@gmail.com
RUBY_VERSION=2.2.3
LESSCLOSE=/usr/bin/lesspipe %s %s
HISTFILE=/home/vagrant/.bash_history
LESS_TERMCAP_se=
NPM_HOME=/home/vagrant/local/node/npm
_=/home/vagrant/.gem/ruby/2.2.3/bin/bundle
_ORIGINAL_GEM_PATH=/home/vagrant/.gem/ruby/2.2.3:/home/vagrant/.rubies/ruby-2.2.3/lib/ruby/gems/2.2.0:/home/vagrant/local/ruby/gems
RUBYLIB=/home/vagrant/.gem/ruby/2.2.3/gems/bundler-1.10.6/lib
```

**Comment by mrbongiolo** [comment]
And here is the `bundle env` log:
```
Environment

    Bundler   1.10.6
    Rubygems  2.4.5.1
    Ruby      2.2.3p173 (2015-08-18 revision 51636) [x86_64-linux]
    GEM_HOME  /home/vagrant/.gem/ruby/2.2.3
    GEM_PATH  /home/vagrant/.gem/ruby/2.2.3:/home/vagrant/.rubies/ruby-2.2.3/lib/ruby/gems/2.2.0:/home/vagrant/local/ruby/gems
    Git       1.9.1

Gemfile

    source 'https://rubygems.org'
    ruby '2.2.3'

    gem 'foreman'
    gem 'roda'
    gem 'rack-indifferent'
    gem 'passenger'#, '~> 4.0'

Gemfile.lock

    GEM
      remote: https://rubygems.org/
      specs:
        foreman (0.78.0)
          thor (~> 0.19.1)
        passenger (5.0.23)
          rack
          rake (>= 0.8.1)
        rack (1.6.4)
        rack-indifferent (1.1.0)
          rack (>= 1.5)
        rake (10.4.2)
        roda (2.9.0)
          rack
        thor (0.19.1)

    PLATFORMS
      ruby

    DEPENDENCIES
      foreman
      passenger
      rack-indifferent
      roda

    BUNDLED WITH
       1.10.6
```

**Comment by OnixGH** [comment]
Thanks for the extended info. Have you tried telling Passenger the correct path (for standalone that would be using `--ruby`)? Between shells, users, environments, nginx, there's a lot that can go wrong (e.g. Passenger 5 also uses a newer nginx for the standalone engine).

[This thread](https://github.com/postmodern/chruby/issues/258) appears somewhat related.

**Comment by mrbongiolo** [comment]
I'll take a look at that. Thanks.

For now I got it working by setting GEM_HOME manually: `bundle exec passenger start --envvar GEM_HOME=/home/vagrant/.gem/ruby/2.2.3`.

Still need to figure out why it's changing or using other values for those env vars. It might be related to the user used for chruby and bundler itself, not sure. Will try to investigate further.

**Comment by FooBarWidget** [comment]
Hi @mrbongiolo, during the investigation of this issue I've come across a bug in Passenger Standalone. It was supposed to turn the `load_shell_envvar` option off by default, but due to a typo it was on by default. That may have an effect on environment variables.

Can you run the following experiment and tell me whether it works for you?

1. Create an Nginx config template: `cp $(passenger-config about resourcesdir)/templates/standalone/config.erb nginx.conf.erb`
2. Edit nginx.conf.erb, look inside the `server` block for "### BEGIN your own configuration options ###".
3. In there, insert this: `passenger_load_shell_envvars off;`
4. Start Passenger as follows: `bundle exec passenger start --nginx-config-template nginx.conf.erb`

Did that work, and if not, what behavior did you see?

**Comment by mrbongiolo** [comment]
@FooBarWidget sorry for the late answer. I just had a chance to test it today with passenger 5.0.23 and your instructions work perfectly.

**Comment by OnixGH** [comment]
The fix is scheduled for 5.0.24
